File: go/sudoku.go
// Sudoku sover in Go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const boardSize = 9 // rows and column size

type board [][]int

func loadBar(step int, totalSteps int, resolution int, width int) {
	if totalSteps/resolution == 0 {
		return
	}
	if step%(totalSteps/resolution) != 0 {
		return
	}
	var ratio = float64(step) / float64(totalSteps)
	var count = int(ratio * float64(width))
	os.Stdout.Write([]byte(fmt.Sprintf("%3d%% [", int(ratio*100))))
	for i := 0; i < count; i++ {
		os.Stdout.Write([]byte("="))
	}
	for i := count; i < width; i++ {
		os.Stdout.Write([]byte(" "))
	}
	os.Stdout.Write([]byte(fmt.Sprintf("]\r")))
	os.Stdout.Sync()
}

// Build a slice of slices to represent the board from a string
func fromStr(boardRepr string) board {
	b := make([][]int, boardSize)
	for i := 0; i < boardSize; i++ {
		b[i] = make([]int, boardSize)
	}

	x, y := 0, 0
	for i := range boardRepr {
		if boardRepr[i] >= '0' && boardRepr[i] <= '9' {
			b[x][y] = int(boardRepr[i]) - 48
			if y == boardSize-1 {
				x += 1 % boardSize
				y = 0
			} else {
				y += 1 % boardSize
			}
		}
	}
	return b
}

func (b *board) get(row, column int) int {
	bslice := [][]int(*b)
	return bslice[row][column]
}

func (b *board) set(row, column, value int) {
	bslice := [][]int(*b)
	bslice[row][column] = value
}

func (b *board) String() string {
	var buf bytes.Buffer
	for i := 0; i < boardSize; i++ {
		for j := 0; j < boardSize; j++ {
			buf.WriteString(fmt.Sprintf("%d ", b.get(i, j)))
		}
		buf.WriteString("\n")
	}
	return buf.String()
}

// Return the next spot where is possible to put a number
func (b *board) nextEmpty() []int {
	for i := 0; i < boardSize; i++ {
		for j := 0; j < boardSize; j++ {
			if b.get(i, j) == 0 {
				return []int{i, j}
			}
		}
	}
	return nil
}

// Return true if the number n can be put in the board at specific spot
func (b *board) canPut(spot []int, n int) bool {
	x, y := spot[0], spot[1]
	for i := 0; i < boardSize; i++ {
		// test line
		if b.get(i, y) == n {
			return false
		}
		// test column
		if b.get(x, i) == n {
			return false
		}
	}
	// test square
	limx, limy := x-(x%3), y-(y%3)
	for i := limx; i < limx+3; i++ {
		for j := limy; j < limy+3; j++ {
			if b.get(i, j) == n {
				return false
			}
		}
	}

	return true
}

// Solve using backtracking
func (b *board) solve() {
	spot := b.nextEmpty()
	if spot == nil {
		return
	}
	x, y := spot[0], spot[1]
	for i := 1; i < 10; i++ {
		if b.canPut(spot, i) {
			b.set(x, y, i)
			b.solve()
			if b.nextEmpty() == nil {
				return
			}
		}
	}
	// solution not found, backtrack
	b.set(x, y, 0)
}

func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func processBatch(input io.Reader, output io.Writer) {
	data, err := ioutil.ReadAll(input)
	check(err)
	var solvedOutput bytes.Buffer
	lines := strings.Split(string(data), "\n")
	total := len(lines)
	count := 0
	before := time.Now()
	for _, line := range lines {
		sudoku := fromStr(line)
		sudoku.solve()
		solvedOutput.WriteString(sudoku.String())
		solvedOutput.WriteString("\n")
		loadBar(count, total, 20, 50)
		count++
	}
	diff := time.Now().Sub(before)
	fmt.Printf("-- Solved %d sudokus. Elapsed time: %f seconds\n", count, diff.Seconds())
	_, err = output.Write(solvedOutput.Bytes())
	check(err)
}

func main() {
	inputFile, err := os.Open(os.Args[1])
	check(err)
	input := bufio.NewReader(inputFile)
	outputFile, err := os.Create("solved_" + filepath.Base(inputFile.Name()))
	check(err)
	output := bufio.NewWriter(outputFile)
	defer inputFile.Close()
	defer outputFile.Close()
	processBatch(input, output)
}

File: cpp/Sudoku.cpp
#include "Sudoku.h"
#include <algorithm>
#include <iostream>
#include <iomanip>
#include <iterator>
#include <vector>

Sudoku::Sudoku(const std::string &data) {
  int i = 0, j = 0;
	for (auto it : data) {
		if (isdigit(it)) {
			board[i][j] = it - 48;
		}
    if (j == BOARD_SIZE - 1) {
      i += 1 % BOARD_SIZE;
      j = 0;
    } else {
      j += 1 % BOARD_SIZE;
    }
	}
}

void Sudoku::solve() {
  int spot[] = {0, 0};
  if (!nextEmpty(spot)) {
		return;
	}

	for (int i=1; i<10; i++) {
		if (canPut(spot, i)) {
			board[spot[0]][spot[1]] = i;
			solve();
			int _[] = {0, 0};
			if (!nextEmpty(_)) {
				return;
			}
		}
	}
  board[spot[0]][spot[1]] = 0; // solution not found, backtrack
}

bool Sudoku::nextEmpty(int spot[2]) {
	for (int i=0; i<BOARD_SIZE; i++) {
		for (int j=0; j<BOARD_SIZE; j++) {
			if (board[i][j] == 0) {
				spot[0] = i;
				spot[1] = j;
				return true;
			}
		}
	}
	return false;
}

bool Sudoku::canPut(const int spot[2], int val) {
	int x = spot[0];
	int y = spot[1];
	for (int i=0; i<BOARD_SIZE; i++) {
    // test row
		if (board[i][y] == val) {
			return false;
		}
    // test column
		if (board[x][i] == val) {
			return false;
		}
	}
  // test square
	int a = x-(x%3);
	int b = y-(y%3);
	for (int i=a; i<a+3; i++) {
		for (int j=b; j<b+3; j++) {
			if (board[i][j] == val) {
				return false;
			}
		}
	}
	return true;
}

std::ostream& operator<<(std::ostream& os, const Sudoku& sudoku) {
  for (int i=0; i<BOARD_SIZE; i++) {
		for (int j=0; j<BOARD_SIZE; j++) {
			os << sudoku.board[i][j] << " ";
		}
		os << "\n";
	}
	os << "\n";
  return os;
}

void progress_bar(int step, int total_steps) {
  int resolution = 100;
  int width = 50;
  if (total_steps / resolution == 0) return;
  double ratio = double(step) / total_steps;
  int count = ratio * width;
  std::string fill(count, '=');
  std::cout << std::setiosflags(std::ios::fixed)
            << std::setprecision(2)
            << ratio * 100.0 << " %"
            << " ["
            << std::setw(width)
            << std::left
            << fill
            << "]"
            << '\r';
}

void Sudoku::process(std::istream& input, std::ostream& output) {
  std::vector<std::string> lines;
  std::copy(std::istream_iterator<std::string>(input),
            std::istream_iterator<std::string>(),
            std::back_inserter(lines));
  int count = 0;
  int total = lines.size();
  for (auto& line: lines) {
    Sudoku sudoku(line);
    sudoku.solve();
    output << sudoku;
    count++;
    progress_bar(count, total);
  }
  std::cout << std::endl;
}
